applications/GeoMechanicsApplication/custom_processes/geo_integration_values_extrapolation_to_nodes_process.cpp

```diff
@@ -0,0 +1,242 @@
+      mpExtrapolator(std::make_unique<NodalExtrapolator>())
+{
+    const Parameters default_parameters = GetDefaultParameters();
+    rParameters.ValidateAndAssignDefaults(default_parameters);
+
+    GetVariableLists(rParameters);
+}
+
+GeoIntegrationValuesExtrapolationToNodesProcess::~GeoIntegrationValuesExtrapolationToNodesProcess() = default;
+
+const Parameters GeoIntegrationValuesExtrapolationToNodesProcess::GetDefaultParameters() const
+{
+    return Parameters(R"(
+    {
+        "model_part_name"            : "",
+        "echo_level"                 : 0,
+        "average_variable"           : "NODAL_AREA",
+        "list_of_variables"          : []
+    })");
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::GetVariableLists(const Parameters& rParameters)
+{
+    for (const std::string& r_variable_name : rParameters["list_of_variables"].GetStringArray()) {
+        KRATOS_ERROR_IF_NOT(TryAddVariableToList(r_variable_name, mDoubleVariables) ||
+                            TryAddVariableToList(r_variable_name, mArrayVariables) ||
+                            TryAddVariableToList(r_variable_name, mVectorVariables) ||
+                            TryAddVariableToList(r_variable_name, mMatrixVariables))
+            << "Only double, array, vector and matrix variables are allowed in the "
+               "variables list."
+            << std::endl;
+    }
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::Execute()
+{
+    ExecuteBeforeSolutionLoop();
+    ExecuteFinalizeSolutionStep();
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::ExecuteBeforeSolutionLoop()
+{
+    InitializeAverageVariablesForElements();
+    InitializeVectorAndMatrixSizesOfVariables();
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::InitializeAverageVariablesForElements() const
+{
+    VariableUtils().SetNonHistoricalVariable(mrAverageVariable, 0.0, mrModelPart.Nodes());
+
+    block_for_each(mrModelPart.Elements(), [this](Element& rElement) {
+        if (rElement.IsActive()) {
+            for (auto& node : rElement.GetGeometry()) {
+                auto& node_var_to_update = node.GetValue(mrAverageVariable);
+                AtomicAdd(node_var_to_update, 1.0);
+            }
+        }
+    });
+
+    mrModelPart.GetCommunicator().AssembleNonHistoricalData(mrAverageVariable);
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::InitializeVectorAndMatrixSizesOfVariables()
+{
+    if (ModelPartContainsAtLeastOneElement()) {
+        auto r_first_element = mrModelPart.Elements().begin();
+
+        const GeometryData::IntegrationMethod this_integration_method =
+            r_first_element->GetIntegrationMethod();
+        const GeometryType::IntegrationPointsArrayType& integration_points =
+            r_first_element->GetGeometry().IntegrationPoints(this_integration_method);
+        const SizeType integration_points_number = integration_points.size();
```

Comment:
If the only thing we're interested in is the number of integration points, I would rewrite this as follows:
```suggestion
        const auto integration_points_number = r_first_element->GetGeometry().IntegrationPointsNumber(r_first_element->GetIntegrationMethod());
```

Reply:
Done